In [1]:
import pandas as pd

In [2]:
train = pd.read_csv("train_imperson_without4n7_balanced_data.csv")

Delete columns that have zero variance

In [3]:
col_to_delete = [k for k in train.columns if train[k].std()==0]

Of which there are

In [4]:
len(col_to_delete)

74

In [5]:
train = train.drop(col_to_delete,axis=1)
trainX = train[train.columns[:-1]]
trainY = train["155"]

In [6]:
trainX0 = trainX[trainY==0]
trainX1 = trainX[trainY>0]

In [7]:
len(trainX)

97044

# Feature selection

Build a data table to hold information regarding the remaining variables

(Some `pandas` operations are easier if the feature name is a regular column but it's nice for that to be the index too)

In [9]:
features = pd.DataFrame(data = trainX.columns, columns = ['feature'], index = train.columns[:-1])

In [10]:
features["mu0"] = features.feature.apply(lambda f: trainX0[f].mean())
features["mu1"] = features.feature.apply(lambda f: trainX1[f].mean())

features["sigma0"] = features.feature.apply(lambda f: trainX0[f].std())
features["sigma1"] = features.feature.apply(lambda f: trainX1[f].std())

features["rho"] = features.feature.apply(lambda f: train[f].corr(train["155"],method='kendall'))

features.head()

feature       mu0       mu1    sigma0    sigma1       rho
5        5  0.005345  0.007160  0.018728  0.011431  0.517106
6        6  0.005345  0.007160  0.018728  0.011431  0.517106
8        8  0.349147  0.038526  0.450527  0.006735 -0.119957
9        9  0.349147  0.038526  0.450527  0.006735 -0.119957
14      14  0.999567  1.000000  0.020799  0.000000  0.014712

## correlation

Sorting features by the absolute value of $\rho$, since a negative correlation is as predictive as a postive one

In [11]:
features["absrho"]=abs(features["rho"])

In [12]:
features.sort_values(by="absrho", ascending=False).loc[features["absrho"]>0.1]

feature           mu0       mu1    sigma0    sigma1       rho    absrho
67       67  7.086695e-01  0.045289  0.260226  0.157740 -0.783199  0.783199
76       76  5.391637e-03  0.001196  0.022788  0.018824 -0.747373  0.747373
79       79  2.451657e-03  0.005380  0.025173  0.001334  0.725743  0.725743
71       71  4.843164e-03  0.676456  0.069425  0.467833  0.708561  0.708561
50       50  4.031573e-01  0.999979  0.490537  0.004540  0.652165  0.652165
51       51  5.964099e-01  0.000021  0.490622  0.004540 -0.651828  0.651828
47       47  5.018176e-01  0.000022  0.450187  0.004542 -0.650500  0.650500
75       75  5.426613e-03  0.000290  0.011442  0.002490 -0.620754  0.620754
68       68  4.490540e-01  0.921829  0.470411  0.268404  0.527840  0.527840
6         6  5.344587e-03  0.007160  0.018728  0.011431  0.517106  0.517106
5         5  5.344587e-03  0.007160  0.018728  0.011431  0.517106  0.517106
73       73  4.862330e-01  0.921870  0.499816  0.268378  0.477183  0.477183
146     146  5.687225e-02  0.000000  0.079131  0.000000 -0.453468  0.453468
145     145  5.687225e-02  0.000000  0.079131  0.000000 -0.453468  0.453468
80       80  1.127750e-03  0.001474  0.008127  0.000000  0.452988  0.452988
38       38  6.591980e-01  0.768314  0.132439  0.024954  0.448580  0.448580
66       66  6.336301e-01  0.921870  0.407407  0.268378  0.420697  0.420697
77       77  3.892950e-02  0.008852  0.095921  0.038162 -0.289967  0.289967
70       70  1.506739e-01  0.000021  0.357734  0.004540 -0.285381  0.285381
78       78  2.422050e-03  0.001731  0.025238  0.000020  0.269324  0.269324
110     110  1.527235e-05  0.000000  0.000041  0.000000 -0.256019  0.256019
140     140  2.261501e-01  0.320148  0.301992  0.199302  0.245990  0.245990
82       82  3.597495e-01  0.091727  0.329202  0.055956 -0.234276  0.234276
142     142  2.310503e-01  0.291817  0.317456  0.187557  0.230346  0.230346
107     107  4.734967e-03  0.000000  0.021979  0.000000 -0.204285  0.204285
122     122  3.645072e-04  0.000000  0.001388  0.000000 -0.193121  0.193121
94       94  5.585095e-02  0.000289  0.229636  0.016984 -0.168196  0.168196
126     126  6.633280e-07  0.000000  0.000003  0.000000 -0.148360  0.148360
130     130  4.307325e-02  0.000000  0.203024  0.000000 -0.148360  0.148360
129     129  6.805573e-07  0.000000  0.000003  0.000000 -0.148360  0.148360
127     127  3.379894e-04  0.000000  0.001593  0.000000 -0.148359  0.148359
128     128  1.284308e-06  0.000000  0.000006  0.000000 -0.148253  0.148253
9         9  3.491472e-01  0.038526  0.450527  0.006735 -0.119957  0.119957
8         8  3.491472e-01  0.038526  0.450527  0.006735 -0.119957  0.119957
61       61  6.077220e-01  0.634935  0.146420  0.074405 -0.106813  0.106813
141     141  7.549768e-03  0.000000  0.051051  0.000000 -0.105829  0.105829
143     143  1.107949e-02  0.000000  0.085167  0.000000 -0.105542  0.105542

## t-tests

t-tests establish the probability $p$ that we see a least that difference in the mean of a feature for the two conditions under the null hypothesis that they come from the same distribution

In [13]:
from scipy.stats import ttest_ind

In [14]:
features["t-test p"] = features.feature.apply(
    lambda f: ttest_ind(trainX0[f],
                        trainX1[f],
                        equal_var=False)[1]
)
features.head()

feature       mu0       mu1    sigma0    sigma1       rho    absrho  \
5        5  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
6        6  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
8        8  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
9        9  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
14      14  0.999567  1.000000  0.020799  0.000000  0.014712  0.014712   

        t-test p  
5   4.460255e-74  
6   4.460255e-74  
8   0.000000e+00  
9   0.000000e+00  
14  4.583503e-06

Sort from lowest probability

In [15]:
features.sort_values(by="t-test p")

feature       mu0       mu1    sigma0    sigma1       rho    absrho  \
154     154  0.334940  0.022008  0.458596  0.006407  0.006752  0.006752   
94       94  0.055851  0.000289  0.229636  0.016984 -0.168196  0.168196   
77       77  0.038930  0.008852  0.095921  0.038162 -0.289967  0.289967   
75       75  0.005427  0.000290  0.011442  0.002490 -0.620754  0.620754   
73       73  0.486233  0.921870  0.499816  0.268378  0.477183  0.477183   
..      ...       ...       ...       ...       ...       ...       ...   
105     105  0.000021  0.000000  0.002780  0.000000 -0.005560  0.005560   
83       83  0.000041  0.000000  0.006420  0.000000 -0.004540  0.004540   
138     138  0.500008  0.500000  0.001725  0.000000 -0.003210  0.003210   
133     133  0.000021  0.000000  0.004540  0.000000 -0.003210  0.003210   
97       97  0.000021  0.000000  0.004540  0.000000 -0.003210  0.003210   

     t-test p  
154  0.000000  
94   0.000000  
77   0.000000  
75   0.000000  
73   0.000000  
..        ...  
105  0.102471  
83   0.157301  
138  0.317315  
133  0.317315  
97   0.317315  

[78 rows x 8 columns]

How many features have $p<0.05$?

In [16]:
len(features[features["t-test p"] < 0.05])

70

$p<0.000000001$?

In [17]:
len(features[features["t-test p"] < 0.000000001])

51

So most of the features have a significantly different mean under the two conditions

## unique values

Some features appear to only have a few values, eg

In [18]:
len(train["133"].unique())

2

Add a column to track the number of unique values a feature takes

In [19]:
features["vals"] = features.feature.apply(lambda f: len(train[f].unique()))

In [20]:
features.sort_values(by="vals").head(40)

feature           mu0       mu1    sigma0    sigma1       rho    absrho  \
129     129  6.805573e-07  0.000000  0.000003  0.000000 -0.148360  0.148360   
106     106  1.482495e-03  0.000018  0.035814  0.003934 -0.028733  0.028733   
71       71  4.843164e-03  0.676456  0.069425  0.467833  0.708561  0.708561   
72       72  1.978484e-03  0.000082  0.044437  0.009079 -0.029548  0.029548   
73       73  4.862330e-01  0.921870  0.499816  0.268378  0.477183  0.477183   
113     113  2.649747e-05  0.000000  0.003370  0.000000 -0.005560  0.005560   
111     111  3.297473e-04  0.000577  0.018156  0.024015  0.005809  0.005809   
133     133  2.060921e-05  0.000000  0.004540  0.000000 -0.003210  0.003210   
83       83  4.121842e-05  0.000000  0.006420  0.000000 -0.004540  0.004540   
84       84  1.236552e-04  0.000000  0.011119  0.000000 -0.007863  0.007863   
86       86  1.648737e-04  0.000000  0.012839  0.000000 -0.009080  0.009080   
89       89  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
90       90  9.636866e-02  0.077305  0.295099  0.267078 -0.033849  0.033849   
93       93  6.535180e-02  0.041012  0.247148  0.198321 -0.054233  0.054233   
94       94  5.585095e-02  0.000289  0.229636  0.016984 -0.168196  0.168196   
97       97  2.060921e-05  0.000000  0.004540  0.000000 -0.003210  0.003210   
98       98  9.408104e-02  0.077305  0.291944  0.267078 -0.029967  0.029967   
108     108  1.250960e-04  0.000118  0.000419  0.000408 -0.008767  0.008767   
118     118  1.021598e-01  0.077965  0.302862  0.268119 -0.042259  0.042259   
121     121  4.253581e-05  0.000000  0.000439  0.000000 -0.068407  0.068407   
70       70  1.506739e-01  0.000021  0.357734  0.004540 -0.285381  0.285381   
138     138  5.000078e-01  0.500000  0.001725  0.000000 -0.003210  0.003210   
126     126  6.633280e-07  0.000000  0.000003  0.000000 -0.148360  0.148360   
14       14  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
15       15  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
16       16  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
18       18  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
20       20  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
26       26  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
29       29  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
62       62  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
130     130  4.307325e-02  0.000000  0.203024  0.000000 -0.148360  0.148360   
51       51  5.964099e-01  0.000021  0.490622  0.004540 -0.651828  0.651828   
43       43  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
123     123  1.236552e-03  0.000000  0.035143  0.000000 -0.024873  0.024873   
50       50  4.031573e-01  0.999979  0.490537  0.004540  0.652165  0.652165   
52       52  9.995672e-01  1.000000  0.020799  0.000000  0.014712  0.014712   
105     105  2.060921e-05  0.000000  0.002780  0.000000 -0.005560  0.005560   
145     145  5.687225e-02  0.000000  0.079131  0.000000 -0.453468  0.453468   
127     127  3.379894e-04  0.000000  0.001593  0.000000 -0.148359  0.148359   

          t-test p  vals  
129   0.000000e+00     2  
106   3.530465e-19     2  
71    0.000000e+00     2  
72    3.419408e-20     2  
73    0.000000e+00     2  
113   8.326452e-02     2  
111   7.037820e-02     2  
133   3.173155e-01     2  
83    1.573013e-01     2  
84    1.430437e-02     2  
86    4.676670e-03     2  
89    4.583503e-06     2  
90    5.217817e-26     2  
93    4.042413e-64     2  
94    0.000000e+00     2  
97    3.173155e-01     2  
98    9.900238e-21     2  
108   6.312788e-03     2  
118   1.302655e-39     2  
121  9.223437e-101     2  
70    0.000000e+00     2  
138   3.173155e-01     2  
126   0.000000e+00     2  
14    4.583503e-06     2  
15    4.583503e-06     2  
16    4.583503e-06     2  
18    4.583503e-06     2  
20    4.

There are a quite a few 2-value features, it might be worth looking at what proportion of the values for a feature are non-zero overall, and under each class

In [21]:
l = len(trainX)
l0 = len(trainX0)
l1 = len(trainX1)
features["nonzero"] = features.feature.apply(lambda f: len(trainX.loc[trainX[f]!=0])/l)
features["nonzero0"] = features.feature.apply(lambda f: len(trainX0.loc[trainX0[f]!=0])/l0)
features["nonzero1"] = features.feature.apply(lambda f: len(trainX1.loc[trainX1[f]!=0])/l1)

In [22]:
features.sort_values(by="nonzero").loc[features["nonzero"]<0.001]

feature       mu0       mu1    sigma0    sigma1       rho    absrho  \
97       97  0.000021  0.000000  0.004540  0.000000 -0.003210  0.003210   
133     133  0.000021  0.000000  0.004540  0.000000 -0.003210  0.003210   
83       83  0.000041  0.000000  0.006420  0.000000 -0.004540  0.004540   
105     105  0.000021  0.000000  0.002780  0.000000 -0.005560  0.005560   
113     113  0.000026  0.000000  0.003370  0.000000 -0.005560  0.005560   
88       88  0.000029  0.000000  0.003466  0.000000 -0.007863  0.007863   
84       84  0.000124  0.000000  0.011119  0.000000 -0.007863  0.007863   
86       86  0.000165  0.000000  0.012839  0.000000 -0.009080  0.009080   
117     117  0.000013  0.000000  0.001288  0.000000 -0.009080  0.009080   
111     111  0.000330  0.000577  0.018156  0.024015  0.005809  0.005809   
123     123  0.001237  0.000000  0.035143  0.000000 -0.024873  0.024873   
144     144  0.000693  0.000000  0.018810  0.000000 -0.028718  0.028718   
106     106  0.001482  0.000018  0.035814  0.003934 -0.028733  0.028733   

         t-test p  vals   nonzero  nonzero0  nonzero1  
97   3.173155e-01     2  0.000010  0.000021  0.000000  
133  3.173155e-01     2  0.000010  0.000021  0.000000  
83   1.573013e-01     2  0.000021  0.000041  0.000000  
105  1.024710e-01     3  0.000031  0.000062  0.000000  
113  8.326452e-02     2  0.000031  0.000062  0.000000  
88   6.551449e-02     4  0.000062  0.000124  0.000000  
84   1.430437e-02     2  0.000062  0.000124  0.000000  
86   4.676670e-03     2  0.000082  0.000165  0.000000  
117  2.637066e-02     6  0.000082  0.000165  0.000000  
111  7.037820e-02     2  0.000453  0.000330  0.000577  
123  9.317263e-15     2  0.000618  0.001237  0.000000  
144  5.116166e-16    75  0.000824  0.001649  0.000000  
106  3.530465e-19     2  0.000866  0.001711  0.000021

These features are extremely sparse, so may be noise.

Which features have good distribution, some correlation, and low p value?

In [23]:
features.loc[(features["nonzero"]>0.01) & (features["absrho"]>0.2) & (features["t-test p"]<0.01)].feature.values


array(['5', '6', '38', '47', '50', '51', '66', '67', '68', '70', '71',
       '73', '75', '76', '77', '78', '79', '80', '82', '107', '110',
       '140', '142', '145', '146'], dtype=object)

Parameterise this for ease of comparison later:

In [24]:
def feature_select(nonzero=0, absrho=0, ttestp=1):
    return features.loc[(features["nonzero"]>=nonzero) & (features["absrho"]>=absrho) & (features["t-test p"]<=ttestp)].feature.values

In [25]:
feature_set1 = feature_select(nonzero=0.01, absrho=0.2, ttestp=0.01)
feature_set1

array(['5', '6', '38', '47', '50', '51', '66', '67', '68', '70', '71',
       '73', '75', '76', '77', '78', '79', '80', '82', '107', '110',
       '140', '142', '145', '146'], dtype=object)

## baselining with logistic regression

Try this set of features in a logistic regression to get a baseline metric for evaluation.

Need to consider what is the most appropriate metric.

Is it more important to detect as many instrusions as possible (recall) or minimise false alarms (precision), or a balance like f1?

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, f1_score, recall_score, precision_score

Split the training data 60/40 into training and validation data

In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX[feature_set1], trainY, test_size=0.4)

In [27]:
# no regularisation at this point
lr = LogisticRegression(solver="sag", penalty='none').fit(X_train,Y_train)

/home/ian/jupyter/env3.6/lib/python3.6/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


In [28]:
confusion_matrix(Y_test,lr.predict(X_test))

array([[19098,   496],
       [  187, 19037]])

In [29]:
def evaluate(model, ytest, xtest):
    pred = model.predict(xtest)
    p = precision_score(ytest,pred)
    r = recall_score(ytest,pred)
    f = f1_score(ytest,pred)
    return p,r,f

In [30]:
print(evaluate(lr, Y_test, X_test))

(0.9746070752060615, 0.9902725759467332, 0.9823773769899632)


In [31]:
def thresholds_to_metrics(nonzero=0, absrho=0, ttestp=1):
    selection = feature_select(nonzero=nonzero, absrho=absrho, ttestp=ttestp)
    X_train, X_test, Y_train, Y_test = train_test_split(trainX[selection], trainY, test_size=0.4)
    lr = LogisticRegression(solver="sag", penalty='none').fit(X_train,Y_train)
    p,r,f = evaluate(lr, Y_test, X_test)
    return p,r,f,len(selection)

In [32]:
# no feature selection
thresholds_to_metrics()

/home/ian/jupyter/env3.6/lib/python3.6/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


(0.9924512904212996, 0.9982556946439565, 0.9953450304363395, 78)

In [33]:
import matplotlib.pyplot as plt

That's 99.8% recall in a simple logistic regression on all non-constant features. This poses a challenge as it does not translate to high recall on the test data.

## PCA

Looks like some features are highly correlated, so perhaps PCA or clustering on features first.

In [27]:
from sklearn.decomposition import PCA

In [28]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX, trainY, test_size=0.4)

In [29]:
pca = PCA(n_components=20)

In [38]:
pca_train = pca.fit_transform(X_train)
pca_test = pca.transform(X_test)

In [39]:
lr = LogisticRegression(solver='sag', penalty='none').fit(pca_train,Y_train)

/home/ian/jupyter/env3.6/lib/python3.6/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


In [40]:
recall_score(Y_test, lr.predict(pca_test))

0.997430362832768

Twenty component PCA does fairly well.

Let's look at a few more or fewer components. We can do that by taking say fifty and then keeping the first 5, 10, 20, 50 of them

In [41]:
pca = PCA(n_components=50)

In [42]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX, trainY, test_size=0.4)

In [43]:
pca_train = pca.fit_transform(X_train)
pca_test = pca.transform(X_test)

In [ ]:
for pcs in [5,10,20,50]:
    lr = LogisticRegression(solver='sag', penalty='none').fit(pca_train[:,:pcs],Y_train)
    score = recall_score(Y_test, lr.predict(pca_test[:,:pcs]))
    print(f"{pcs}:\t{score}")

## Mutual information

In [30]:
from sklearn.feature_selection import mutual_info_classif

Consider any feature with fewer than 10 unique values to be discrete. Is that reasonable?

In [31]:
mi = mutual_info_classif(trainX,trainY, discrete_features=(features.vals<10))
mi

array([4.46383568e-01, 4.46986286e-01, 6.33835575e-01, 6.33287379e-01,
       1.50018172e-04, 1.50018172e-04, 1.50018172e-04, 1.50018172e-04,
       1.50018172e-04, 1.50018172e-04, 1.50018172e-04, 6.48616928e-01,
       1.50018172e-04, 3.06428902e-01, 7.84877466e-03, 2.72272530e-01,
       2.72002744e-01, 1.50018172e-04, 3.92032401e-01, 1.50018172e-04,
       5.37829426e-01, 1.43635424e-01, 4.69867231e-01, 1.75160966e-01,
       5.51963961e-02, 3.11386227e-01, 5.41650770e-04, 1.23919498e-01,
       2.73477949e-01, 4.77187659e-01, 4.86752325e-01, 2.06630069e-01,
       5.26412831e-01, 1.54276462e-01, 5.81413356e-01, 1.42854269e-05,
       4.28575551e-05, 5.71442564e-05, 4.28575551e-05, 1.50018172e-04,
       5.73936203e-04, 1.48303711e-03, 1.89492770e-02, 7.14266037e-06,
       4.49655821e-04, 2.47054140e-03, 2.14282997e-05, 5.44435479e-04,
       3.02999194e-02, 3.84346475e-05, 2.09096746e-03, 4.49100997e-02,
       1.70852413e-05, 2.64472542e-03, 2.14282997e-05, 5.71442564e-05,
      

In [33]:
features["mi"] = mi
features.sort_values(by="mi", ascending=False).head(30)

feature       mu0       mu1    sigma0    sigma1       rho    absrho  \
38       38  0.659198  0.768314  0.132439  0.024954  0.448580  0.448580   
8         8  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
9         9  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
82       82  0.359749  0.091727  0.329202  0.055956 -0.234276  0.234276   
142     142  0.231050  0.291817  0.317456  0.187557  0.230346  0.230346   
140     140  0.226150  0.320148  0.301992  0.199302  0.245990  0.245990   
64       64  0.670202  0.683830  0.273232  0.149417 -0.019666  0.019666   
79       79  0.002452  0.005380  0.025173  0.001334  0.725743  0.725743   
154     154  0.334940  0.022008  0.458596  0.006407  0.006752  0.006752   
77       77  0.038930  0.008852  0.095921  0.038162 -0.289967  0.289967   
76       76  0.005392  0.001196  0.022788  0.018824 -0.747373  0.747373   
67       67  0.708669  0.045289  0.260226  0.157740 -0.783199  0.783199   
6         6  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
5         5  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
61       61  0.607722  0.634935  0.146420  0.074405 -0.106813  0.106813   
71       71  0.004843  0.676456  0.069425  0.467833  0.708561  0.708561   
47       47  0.501818  0.000022  0.450187  0.004542 -0.650500  0.650500   
75       75  0.005427  0.000290  0.011442  0.002490 -0.620754  0.620754   
50       50  0.403157  0.999979  0.490537  0.004540  0.652165  0.652165   
51       51  0.596410  0.000021  0.490622  0.004540 -0.651828  0.651828   
78       78  0.002422  0.001731  0.025238  0.000020  0.269324  0.269324   
68       68  0.449054  0.921829  0.470411  0.268404  0.527840  0.527840   
80       80  0.001128  0.001474  0.008127  0.000000  0.452988  0.452988   
66       66  0.633630  0.921870  0.407407  0.268378  0.420697  0.420697   
146     146  0.056872  0.000000  0.079131  0.000000 -0.453468  0.453468   
145     145  0.056872  0.000000  0.079131  0.000000 -0.453468  0.453468   
73       73  0.486233  0.921870  0.499816  0.268378  0.477183  0.477183   
70       70  0.150674  0.000021  0.357734  0.004540 -0.285381  0.285381   
110     110  0.000015  0.000000  0.000041  0.000000 -0.256019  0.256019   
107     107  0.004735  0.000000  0.021979  0.000000 -0.204285  0.204285   

          t-test p   vals   nonzero  nonzero0  nonzero1        mi  
38    0.000000e+00  25023  0.999784  0.999567  1.000000  0.648617  
8     0.000000e+00    399  0.868575  0.737150  1.000000  0.633836  
9     0.000000e+00    399  0.868575  0.737150  1.000000  0.633287  
82    0.000000e+00   4096  0.867421  0.734842  1.000000  0.581413  
142  3.517662e-286  20359  0.692150  0.462429  0.921870  0.574457  
140   0.000000e+00  17320  0.692150  0.462429  0.921870  0.574385  
64    5.596116e-22     20  0.996476  0.992952  1.000000  0.537829  
79   1.914366e-143     46  0.867668  0.735337  1.000000  0.526413  
154   0.000000e+00    333  0.705793  0.489716  0.921870  0.522978  
77    0.000000e+00     84  0.867668  0.735337  1.000000  0.486752  
76   1.633120e-213    133  0.999681  0.999361  1.000000  0.477188  
67    0.000000e+00     12  0.509305  0.940481  0.078130  0.469867  
6     4.460255e-74   8718  1.000000  1.000000  1.000000  0.446986  
5     4.460255e-74   8718  1.000000  1.000000  1.000000  0.446384  
61   5.360185e-289     63  0.999918  0.999918  0.999918  0.392032  
71    0.000000e+00      2  0.340650  0.004843  0.676456  0.311386  
47    0.000000e+00     12  0.323863  0.647644  0.000082  0.306429  
75    0.000000e+00     80  0.323183  0.632929  0.013437  0.273478  
50    0.000000e+00      2  0.701568  0.403157  0.999979  0.272273  
51    0.000000e+00      2  0.298215  0.596410  0.000021  0.272003  
78    1.662203e-09     42  0.868575  0.737150  1.000000  0.206630  
68    0.000000e+00      3  0.711584  0.501298  0.921870  0.175161  
80    5.789829e-21     18  0.867988  0.735975  1.000000  0.154276  
66    0.000000e+00      3  0.843916  0.765962 

Some of the features with relatively high mutual information with the target have low correlation which suggests that these features have a relation with the target if not a *linear* one.

We can select features with high linear correlation *or* mutual information

In [34]:
features.loc[(features.absrho > 0.5)|(features.mi > 0.3)]

feature       mu0       mu1    sigma0    sigma1       rho    absrho  \
5         5  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
6         6  0.005345  0.007160  0.018728  0.011431  0.517106  0.517106   
8         8  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
9         9  0.349147  0.038526  0.450527  0.006735 -0.119957  0.119957   
38       38  0.659198  0.768314  0.132439  0.024954  0.448580  0.448580   
47       47  0.501818  0.000022  0.450187  0.004542 -0.650500  0.650500   
50       50  0.403157  0.999979  0.490537  0.004540  0.652165  0.652165   
51       51  0.596410  0.000021  0.490622  0.004540 -0.651828  0.651828   
61       61  0.607722  0.634935  0.146420  0.074405 -0.106813  0.106813   
64       64  0.670202  0.683830  0.273232  0.149417 -0.019666  0.019666   
67       67  0.708669  0.045289  0.260226  0.157740 -0.783199  0.783199   
68       68  0.449054  0.921829  0.470411  0.268404  0.527840  0.527840   
71       71  0.004843  0.676456  0.069425  0.467833  0.708561  0.708561   
75       75  0.005427  0.000290  0.011442  0.002490 -0.620754  0.620754   
76       76  0.005392  0.001196  0.022788  0.018824 -0.747373  0.747373   
77       77  0.038930  0.008852  0.095921  0.038162 -0.289967  0.289967   
79       79  0.002452  0.005380  0.025173  0.001334  0.725743  0.725743   
82       82  0.359749  0.091727  0.329202  0.055956 -0.234276  0.234276   
140     140  0.226150  0.320148  0.301992  0.199302  0.245990  0.245990   
142     142  0.231050  0.291817  0.317456  0.187557  0.230346  0.230346   
154     154  0.334940  0.022008  0.458596  0.006407  0.006752  0.006752   

          t-test p   vals   nonzero  nonzero0  nonzero1        mi  
5     4.460255e-74   8718  1.000000  1.000000  1.000000  0.446384  
6     4.460255e-74   8718  1.000000  1.000000  1.000000  0.446986  
8     0.000000e+00    399  0.868575  0.737150  1.000000  0.633836  
9     0.000000e+00    399  0.868575  0.737150  1.000000  0.633287  
38    0.000000e+00  25023  0.999784  0.999567  1.000000  0.648617  
47    0.000000e+00     12  0.323863  0.647644  0.000082  0.306429  
50    0.000000e+00      2  0.701568  0.403157  0.999979  0.272273  
51    0.000000e+00      2  0.298215  0.596410  0.000021  0.272003  
61   5.360185e-289     63  0.999918  0.999918  0.999918  0.392032  
64    5.596116e-22     20  0.996476  0.992952  1.000000  0.537829  
67    0.000000e+00     12  0.509305  0.940481  0.078130  0.469867  
68    0.000000e+00      3  0.711584  0.501298  0.921870  0.175161  
71    0.000000e+00      2  0.340650  0.004843  0.676456  0.311386  
75    0.000000e+00     80  0.323183  0.632929  0.013437  0.273478  
76   1.633120e-213    133  0.999681  0.999361  1.000000  0.477188  
77    0.000000e+00     84  0.867668  0.735337  1.000000  0.486752  
79   1.914366e-143     46  0.867668  0.735337  1.000000  0.526413  
82    0.000000e+00   4096  0.867421  0.734842  1.000000  0.581413  
140   0.000000e+00  17320  0.692150  0.462429  0.921870  0.574385  
142  3.517662e-286  20359  0.692150  0.462429  0.921870  0.574457  
154   0.000000e+00    333  0.705793  0.489716  0.921870  0.522978

## Combination selection

In [35]:
feature_set3 = features.loc[(features.absrho > 0.5)|(features.mi > 0.3)].feature
len(feature_set3)

21

In [36]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX[feature_set3], trainY, test_size=0.4)

In [64]:
lr = LogisticRegression(solver="sag", penalty='none').fit(X_train,Y_train)

/home/ian/jupyter/env3.6/lib/python3.6/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


In [65]:
confusion_matrix(Y_test,lr.predict(X_test))

array([[19146,   326],
       [  218, 19128]])

In [68]:
recall_score(Y_test, lr.predict(X_test))

0.988731520727799

Following Tim's lead:

In [54]:
from sklearn.neural_network import MLPClassifier

In [70]:
mlp = MLPClassifier(hidden_layer_sizes=(5, 5), max_iter=1000) #two layers of 5 neurons, 1000 of backprop
mlp.fit(X_train, Y_train.values.ravel()) #train the algo

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 5), learning_rate='constant',
              learning_rate_init=0.001, max_iter=1000, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=None, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [71]:
confusion_matrix(Y_test, mlp.predict(X_test))

array([[19380,    92],
       [   69, 19277]])

So, these 20 offer good recall and precision

Being slightly more selective

In [73]:
feature_set4 = features.loc[(features.absrho > 0.6)|(features.mi > 0.5)].feature
len(feature_set4)

14

In [74]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX[feature_set4], trainY, test_size=0.4)

In [75]:
mlp = MLPClassifier(hidden_layer_sizes=(5, 5), max_iter=1000) #two layers of 5 neurons, 1000 of backprop
mlp.fit(X_train, Y_train.values.ravel()) #train the algo

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 5), learning_rate='constant',
              learning_rate_init=0.001, max_iter=1000, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=None, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [76]:
confusion_matrix(Y_test, mlp.predict(X_test))

array([[19397,    12],
       [ 1547, 17862]])

In [78]:
recall_score(Y_test, mlp.predict(X_test))

0.9202947086403215

In [79]:
precision_score(Y_test, mlp.predict(X_test))

0.9993286337697214

Good precision but recall has suffered slightly

How about a hybrid selection: the top 5 by absolute rho, top 5 by mutual information, and the first ten principle components?

In [37]:
top5rho = features.sort_values(by="absrho", ascending=False).head(5).feature
top5rho

67    67
76    76
79    79
71    71
50    50
Name: feature, dtype: object

In [46]:
top5mi = features.sort_values(by="mi", ascending=False).head(5).feature
top5mi

38      38
8        8
9        9
82      82
142    142
Name: feature, dtype: object

No intersection - mutual information measures different kind of relationship to linear correlation

In [47]:
top10 = pd.concat([top5rho,top5mi]).values

In [48]:
pca = PCA(n_components=10)

In [49]:
pca_train = pca.fit_transform(trainX)
pca_train.shape

(97044, 10)

In [50]:
trainX20 = pd.concat([trainX[top10], pd.DataFrame(pca_train)], axis=1)

In [51]:
trainX20

67        76        79  71  50       38         8         9  \
0      0.30769  0.001127  0.001440   1   0  0.36865  0.009150  0.009150   
1      1.00000  0.002253  0.000000   0   0  0.36867  0.000000  0.000000   
2      0.61538  0.000563  0.001728   0   1  0.36871  0.070588  0.070588   
3      0.61538  0.000563  0.000864   0   1  0.36876  0.094771  0.094771   
4      0.61538  0.000563  0.001728   0   1  0.36880  0.070588  0.070588   
...        ...       ...       ...  ..  ..      ...       ...       ...   
97039  0.61538  0.000563  0.001728   0   1  0.93345  0.038562  0.038562   
97040  0.61538  0.000563  0.001728   0   1  0.93348  0.038562  0.038562   
97041  0.61538  0.000563  0.001728   0   1  0.93350  0.038562  0.038562   
97042  0.61538  0.000563  0.001728   0   1  0.93352  0.038562  0.038562   
97043  0.61538  0.000563  0.001728   0   1  0.93354  0.038562  0.038562   

             82  142         0         1         2         3         4  \
0      0.946280  0.0  0.420522  0.453253 -0.547955 -0.854962  0.586509   
1      0.000000  0.0  0.058966  1.319915 -0.917786 -0.579474 -0.039739   
2      0.255430  0.0 -1.473498  1.431104  1.153064 -0.331561 -0.054824   
3      0.072772  0.0 -1.571292  1.509385  1.416091 -0.455870 -0.079367   
4      0.256900  0.0 -1.471950  1.430648  1.152154 -0.332272 -0.054604   
...         ...  ...       ...       ...       ...       ...       ...   
97039  0.217830  0.0 -1.515816  1.393979  1.143533 -0.350176 -0.095881   
97040  0.218320  0.0 -1.515666  1.393976  1.143294 -0.349872 -0.095791   
97041  0.218560  0.0 -1.515786  1.394070  1.143764 -0.350043 -0.095825   
97042  0.218800  0.0 -1.515541  1.393983  1.143125 -0.349609 -0.095713   
97043  0.219050  0.0 -1.515511  1.394004  1.143144 -0.349531 -0.095686   

              5         6         7         8         9  
0      0.763728  0.569430 -0.064739 -0.041498 -0.076144  
1     -0.049850 -0.026646  0.080835  0.035814  0.052931  
2     -0.045797 -0.225876  0.135513 -0.365991 -0.246507  
3     -0.057258 -0.057565  0.035537 -0.038628  1.115491  
4     -0.044002 -0.229213  0.138189 -0.369595 -0.256010  
...         ...       ...       ...       ...       ...  
97039 -0.076009 -0.229035  0.148365 -0.346684 -0.348898  
97040 -0.075774 -0.228729  0.148165 -0.346722 -0.348996  
97041 -0.075554 -0.228660  0.148134 -0.346823 -0.349008  
97042 -0.075533 -0.228438  0.147977 -0.346769 -0.349090  
97043 -0.075389 -0.228301  0.147891 -0.346808 -0.349132  

[97044 rows x 20 columns]

In [52]:
X_train, X_test, Y_train, Y_test = train_test_split(trainX20, trainY, test_size=0.4)

In [55]:
mlp = MLPClassifier(hidden_layer_sizes=(5, 5), max_iter=1000)
mlp.fit(X_train, Y_train.values.ravel())

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 5), learning_rate='constant',
              learning_rate_init=0.001, max_iter=1000, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=None, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [56]:
confusion_matrix(Y_test, mlp.predict(X_test))

array([[19390,    39],
       [   23, 19366]])

In [57]:
recall_score(Y_test, mlp.predict(X_test))

0.9988137603795967

In [58]:
precision_score(Y_test, mlp.predict(X_test))

0.9979902087090956

So these do well on our hold out set, but on the given test set they perform really badly - ~50% recall...

Export this feature set with the target column to csv

In [65]:
pd.concat([trainX20,trainY], axis=1).to_csv("selection01.csv")